tests/integration/test_shared_db.py

```diff
@@ -0,0 +1,253 @@
+
+        # Deploy and test the first deployment of keystone
+        await deploy_and_relate_keystone_with_mysql(ops_test, KEYSTONE_APP_NAME, 2)
+        await check_successful_keystone_migration(ops_test, server_config_credentials)
+
+        keystone_users = []
+        for unit in ops_test.model.applications[KEYSTONE_APP_NAME].units:
+            unit_address = await unit.get_public_address()
+
+            keystone_users.append(f"keystone@{unit_address}")
+
+        await check_keystone_users_existence(
+            ops_test, server_config_credentials, keystone_users, []
+        )
+
+        # Deploy and test another deployment of keystone
+        await deploy_and_relate_keystone_with_mysql(ops_test, ANOTHER_KEYSTONE_APP_NAME, 2)
+        await check_successful_keystone_migration(ops_test, server_config_credentials)
+
+        another_keystone_users = []
+        for unit in ops_test.model.applications[ANOTHER_KEYSTONE_APP_NAME].units:
+            unit_address = await unit.get_public_address()
+
+            another_keystone_users.append(f"keystone@{unit_address}")
+
+        await check_keystone_users_existence(
+            ops_test, server_config_credentials, keystone_users + another_keystone_users, []
+        )
+
+        # Scale down the second deployment of keystone and confirm that the first deployment
+        # is still active
+        await scale_application(ops_test, ANOTHER_KEYSTONE_APP_NAME, 0)
+        await ops_test.model.remove_application(ANOTHER_KEYSTONE_APP_NAME, block_until_done=True)
+
+        await check_keystone_users_existence(
+            ops_test, server_config_credentials, keystone_users, another_keystone_users
+        )
+
+        # Scale down the primary unit of mysql
+        primary_unit = await get_primary_unit(
+            ops_test,
+            random_unit,
+            APP_NAME,
+            CLUSTER_NAME,
+            server_config_credentials["username"],
+            server_config_credentials["password"],
+        )
+        primary_unit_name = primary_unit.name
+
+        await ops_test.model.destroy_units(primary_unit_name)
+
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == 2
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+
+        await check_keystone_users_existence(
+            ops_test, server_config_credentials, keystone_users, another_keystone_users
+        )
+
+        # Scale mysql back up to 3 units
+        await scale_application(ops_test, APP_NAME, 3)
+
+        # Scale down the first deployment of keystone
+        await scale_application(ops_test, KEYSTONE_APP_NAME, 0)
+        await ops_test.model.remove_application(KEYSTONE_APP_NAME, block_until_done=True)
+
+        # Scale down the mysql application
```

Comment:
ops-test creates a separate model for each test file, which is deleted after the run, so you don't have to run these cleanup steps unless you're planning to add more tests later on.

Reply:
This is useful to catch any error in the departing of units - there's some activity in the relation management tied to it.

Reply:
Ah, I didn't think of that!